### ISSUE TITLE
Editor is rendering using save(), not edit() on custom block

### ISSUE THREAD

**Issue by QSSRobert** [WordPress/gutenberg#12630]
I have a relatively simple custom list block for adding tile links to a page. Adding a new block works as expected, but after saving the page and re-opening to edit later, the editor is rendering the output of the `save` method, not `edit`. This results in a broken editor experience.

**To Reproduce**
Here's my component:
```
registerBlockType("mysite/subpage-tiles-block", {
  title: __("Subpage Tiles"),
  icon: "editor-insertmore", 
  category: "common", 
  keywords: [__("Subpage Tiles")],

  attributes: {
    tiles: {
      source: "query",
      default: [],
      selector: 'li',
      query: {
        index: {
          source: "text",
          selector: "span.tileIndex"
        },
        link: {
          source: "attribute",
          selector: "a",
          attribute: "href"
        },
        image: {
          source: "attribute",
          selector: "img",
          attribute: "src"
        },
        title: {
          source: "text",
          selector: "h2.section-title"
        },
        text: {
          source: "text"
        }
      }
    }
  },

  edit: props => {
    const { tiles } = props.attributes;

    const tilesList = tiles.map(function(tile) {
      return (
        <div>
          <p>
            <span className="removeTile" onClick={() => {
              const newTilesList = tilesList.filter(i => i.index != tile.index ).map(t => {
                if (t.index > tile.index ) {
                  t.index -= 1;
                }
                return t;
              });
              props.setAttributes({
                tilesList: newTilesList
              });
            }}>&times;</span>
          </p>
          <div className="editTile">
            <div>
              <MediaUpload
                onSelect={media => {
                  const image = media.sizes.medium
                    ? media.sizes.medium.url
                    : media.url;
                  const newObject = Object.assign({}, tile, {
                    image: image
                  });
                  props.setAttributes({
                    tiles: [
                      ...tiles.filter(
                        item => item.index != tile.index
                      ),
                      newObject
                    ]
                  });
                }}
                type="image"
                value={tile.image}
                render={({ open }) =>
                  !!tile.image ? (
                    <div>
                      {props.isSelected && (
                        <div className="gts__picture__actions">
                          <a
                            href="#"
                            onClick={() => {
                              const newObject = Object.assign(
                                {},
                                tile,
                                {
                                  image: null
                                }
                              );
                              props.setAttributes({
                                tilesList: [
                                  ...tilesList.filter(
                                    item => item.index != tile.index
                                  ),
                                  newObject
                                ]
                              });
                            }}
                          >
                            × Remove
                          </a>
                        </div>
                      )}

                      <div
                        className="gts__picture__image"
                        style={{
                          backgroundImage: `url(${tile.image})`
                        }}
                        onClick={open}
                      />
                    </div>
                  ) : (
                    <a
                      href="#"
                      className="gts__picture__image"
                      onClick={open}
                    >
                      Select Image
                    </a>
                  )
                }
              />
            </div>
            <div>
              <PlainText
                className="text-input"
                placeholder="page-url/"
                value={tile.link}
                autoFocus
                onChange={link => {
                  const newObject = Object.assign({}, tile, {
                    link: link
                  });
                  props.setAttributes({
                    tiles: [
                      ...tiles.filter(
                        item => item.index != tile.index
                      ),
                      newObject
                    ]
                  });
                }}
              />
            </div>
            <div>
              <PlainText
                className="text-input"
                placeholder="Title"
                value={tile.title}
                onChange={title => {
                  const newObject = Object.assign({}, tile, {
                    title: title
                  });
                  props.setAttributes({
                    tiles: [
                      ...tiles.filter(
                        item => item.index != tile.index
                      ),
                      newObject
                    ]
                  });
                }}
              />
            </div>
            <div>
              <PlainText
                className="text-input"
                placeholder="Description"
                value={tile.text}
                onChange={text => {
                  const newObject = Object.assign({}, tile, {
                    text: text
                  });
                  props.setAttributes({
                    tiles: [
                      ...tiles.filter(
                        item => item.index != tile.index
                      ),
                      newObject
                    ]
                  });
                }}
              />
            </div>
          </div>
        </div>
      );
    });

    return (
      <div className="subpageTilesEditor">
          { tilesList }
        <button className="addTile" onClick={content => props.setAttributes({
          tiles: [
            ...props.attributes.tiles,
            {
              index: props.attributes.tiles.length,
              title: "",
              text: "",
              link: ""
            }
          ]
        })}><span>+</span></button>
      </div>
    );
  },

  save: props => {
    const { tiles } = props.attributes;
    
    const tilesList = tiles.map(function(tile) {
      const { link, image, title, text } = tile;
      const description = text ? <p>{text}</p> : null;
      return (
        <li>
          <a href={link} className="tileLink">
            <div className="image">><img src={image} alt={title} /></div>
            <div className="text">
              <h2 className="section-title">{title}</h2>
             { description }
            </div>
          </a>
          <a href={link} className="crosslink">More</a>
        </li>
      );
    });
    if (tilesList.length) {
      return (
        <ul className="subpage-tiles">
          {tilesList}
        </ul>
      )
    } 
    return (
      <p>No Subpage Tiles</p>
    );
  }
});
```

**Expected behavior**
The editor should render the HTML from the `edit` method, giving a consistent interface.

**Screenshots**
The output of the `edit` method when adding new block
![image](https://user-images.githubusercontent.com/21097139/49558084-37470780-f94d-11e8-81d2-41a23c78d3eb.png)

vs
The output from the `save` method when editing pre-existing blocks
![image](https://user-images.githubusercontent.com/21097139/49558230-a91f5100-f94d-11e8-9975-a6e79e0bc8b4.png)

**Desktop:**
 - OS: Windows 10
 - Chrome

**Additional context**
- Gutenberg 4.6.1

**Comment by youknowriad** [comment]
Looks like this has been answered. Thanks

**Comment by QSSRobert** [comment]
Thank you @talldan for your help!
(and sorry for my delayed reply... I was overseas for the holidays)

I did indeed have an extra `>` in my code, and I added a key to the array elements as you pointed out.

However the bug I was having seems to be caused by improperly initialized attributes. Specifically, this line:

```text: {
          source: "text"
        }```

Adding a `selector: 'p'` fixed the issue.

As Gutenberg tries to parse and rebuild the original block from saved content, the missing selector was causing it to read in and concatenate static text from the `<a>` element following the description.

Now I'm off to carefully read the [Attributes documentation](https://wordpress.org/gutenberg/handbook/designers-developers/developers/block-api/block-attributes/).

Thanks again!
